File: Exercises/Python/machine-learning-ex3/lrCostFunction.py
from __future__ import division
from scipy.io import loadmat
from scipy import linalg, sparse #for linalg
import matplotlib.pyplot as plt #for the graph
import numpy as np #for lin alg operations
from scipy.special import expit #vectorized sigmoid function



datafile = 'ex3data1.mat'
mat = loadmat( datafile )
y=np.mat([[1], [0], [1], [0], [1]])


X=np.mat([[1, .1, .6, 1.1], [1, .2,.7,1.2], [1, .3, .8, 1.3], [1,.4,.9,1.4], [1, .5, 1, 1.5]])
theta_t=np.mat([[-2], [-1], [1], [2]])

m = y.shape[0]

J=0


def sigmoid(z):  
    return 1 / (1 + np.exp(-z));

argument=(np.dot(theta_t.T, X.T)).T
g = sigmoid(argument)

def gradientFunc(theta, X, y, mylambda):
	temp=theta
	temp[0]=0
	grad=(1/m)*(np.dot(X.T,g-y))+(mylambda/m)*temp
	return grad

def costFunc(theta , X, y, mylambda):
    m = X.shape[0]
    summation = sum(np.dot((-1*y).T,np.log(g))-np.dot((1-y).T, np.log(1-g)))
    summation=np.unwrap(summation)[0][0]
    third_term=sum(np.square(theta[1:,:]))
    J=((1./m) * summation)+(mylambda/(2*m))*np.unwrap(third_term)[0][0]
    #(mylambda/(2*m)) is not working
    return J

cost=costFunc(theta_t, X, y, 3);
grad=gradientFunc(theta_t, X, y, 3);

print("Cost at initial theta (zeros): " +  str(cost));
print("Expected cost: 2.534819\n");
print("Gradient at initial theta (zeros): \n");
print(grad);
print("Expected gradients (approx): 0.146561\n -0.548558\n 0.724722\n 1.398003\n");

File: Exercises/R/machine-learning-ex3/lrCostFunction.R
install.packages("R.matlab")
install.packages("sigmoid")
library(R.matlab)
library(sigmoid)

pathname<-"/Users/cmr/Desktop/Machine Learning/Class/Octave/machine-learning-ex3/ex3/ex3data1.mat"
data <- readMat(pathname)

theta_t <- matrix(c(-2, -1, 1, 2), ncol = 1)
X_t <-matrix(c(1, 1, 1, 1, 1, .1, .2, .3, .4, .5, .6 ,.7,.8,.9,1, 1.1, 1.2,1.3,1.4,1.5), ncol=4);
y_t <- matrix(c(1,0,1,0,1), ncol = 1)
lambda_t=3


lrCostFunction <- function(matrixA, matrixB, theta_matrix, lambda){
	m<-length(matrixB);
	length_of_theta <- length(theta_matrix)[1];
	argument<-t(t(theta_matrix)%*%t(matrixA));  
	g<-sigmoid(argument);

	#regularized cost
	sumation<-sum((t(-1*matrixB)%*%log(g))-(t(1-matrixB)%*%log(1-g)));
	J<-((1/m)*(sumation))+(lambda/(2*m))%*%sum((theta_matrix[2:length_of_theta,]^2));

	#regularized gradient
	temp<-theta_matrix;
	temp[1]<-0;
	grad<-(1/m)*(t(matrixA)%*%(g-matrixB))+(lambda/m)*temp;
	newList <- list("cost" = J, "theta" = grad);
	return (newList);
}


returnedList <- lrCostFunction(X_t, y_t, theta_t, lambda_t);

J<-returnedList$cost
grad<-returnedList$theta

cat("Cost: ", J);
print("Expected cost: 2.534819");
cat("Gradients: ", grad);
print("Expected gradients: 0.146561 -0.548558 0.724722 1.398003");
